File: core/CoinChange/CoinChange.py
from typing import List


def coinChange(coins: List[int], amount: int) -> int:
    remain = {0: 0}
    for index in range(1, amount + 1):
        mid = -1
        for coin in coins:
            if index - coin >= 0 and (index - coin) in remain:
                mid = remain[index - coin] if mid == -1 else min(mid, remain[index - coin])
        if mid != -1:
            remain[index] = mid + 1

    return remain[amount] if amount in remain else -1


if __name__ == '__main__':
    print(coinChange([1, 2, 5], 11))

File: core/CoinChange/CoinChange.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;

public class CoinChange {
    public int coinChange(int[] coins, int amount) {
        Map<Integer, Integer> remain = new HashMap<Integer, Integer>(){{
            put(0, 0);
        }};
        for (int i = 1; i <= amount; i++) {
            int mid = -1;
            for (int coin : coins) {
                if (i - coin >= 0 && remain.containsKey(i - coin)) {
                    mid = mid == -1 ? remain.get(i - coin) : Math.min(mid, remain.get(i - coin));
                }
            }
            if (mid != -1) remain.put(i, mid + 1);
        }
        return remain.getOrDefault(amount, -1);
    }
}
